### PR TITLE
IDP Api changes for Extensible User Matching OKTA-93061

### PR DESCRIPTION
**Pull request by ashishgupta-okta**
IDP api changes for Extensible user matching

Primary Reviewers: @larsjohansen-okta @mystiberry-okta 
Reviewers: @federations-okta

### CODE REVIEW

**Comment by mystiberry-okta** [comment]
Looks great with my minor corrections for case. One dumb question--they can use custom profile attributes as well as Okta-defined ones, yes?

File: _source/docs/api/resources/idps.md [hunk 1]
```diff
@@ -1016,9 +1016,12 @@ Specifies the behavior for establishing, validating, and matching a username for
...
+| matchType        | Determines the Okta user profile attribute match conditions for account linking and authentication of the transformed IdP username  | `USERNAME`, `EMAIL`, `USERNAME_OR_EMAIL` or any valid Okta user profile attribute (for example, 'firstName') if you have Inbound SAML Extensible User Matching enabled      | FALSE    | FALSE    |           |           |                                                                     |
 |------------------+-------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------+----------+----------+-----------+-----------+---------------------------------------------------------------------|
 
+> Inbound SAML Extensible User Matching is an Early Access feature which you can request from Okta Support.
+It's been tested to the same level as Generally Available (GA) features, so you  can use it in a production environment.
```
Comment (cwu-okta) [inline _source/docs/api/resources/idps.md]: @mystiberry-okta do we really need to have all this additional verbiage around GA? It becomes repetitive to always add "It's been tested..." for each EA feature. In some places on the developer site we just say the first sentence. Can we do that here ?

@ashishgupta-okta FYI we might want to say in the EA blurb that USERNAME_OR_EMAIL is not supported for extensible user matching.

File: _source/docs/api/resources/idps.md [hunk 2]
```diff
@@ -1016,9 +1016,12 @@ Specifies the behavior for establishing, validating, and matching a username for
...
 | filter           | Optional [regular expression pattern](https://en.wikipedia.org/wiki/Regular_expression) used to filter untrusted IdP usernames      | String                                                 | TRUE     | FALSE    | 0         | 1024      |                                                                     |
-| matchType        | Determines the Okta user profile attribute match conditions for account linking and authentication of the transformed IdP username  | `USERNAME`, `EMAIL`, or `USERNAME_OR_EMAIL`            | FALSE    | FALSE    |           |           |                                                                     |
+| matchType        | Determines the Okta user profile attribute match conditions for account linking and authentication of the transformed IdP username  | `USERNAME`, `EMAIL`, `USERNAME_OR_EMAIL` or any valid Okta user profile attribute (for example, 'firstName') if you have Inbound SAML Extensible User Matching enabled      | FALSE    | FALSE    |           |           |                                                                     |
```
Comment (cwu-okta) [inline _source/docs/api/resources/idps.md]: What about changing this to "...or any valid Okta user profile attribute*." The star would then be similar to a footnote but the footnote text is in the callout for EA. It feels a little wordy to call out a feature directly in the reference doc.

@mystiberry-okta thoughts?

File: _source/docs/api/resources/idps.md (hunk 1 above)
Comment (ashishgupta-okta) [inline _source/docs/api/resources/idps.md]: will change for  USERNAME_OR_EMAIL.

**Comment by ashishgupta-okta** [comment]
@mystiberry-okta Yes, custom profile properties can also be used.

File: _source/docs/api/resources/idps.md (hunk 1 above)
Comment (mystiberry-okta) [inline _source/docs/api/resources/idps.md]: @cwu-okta will remove.

File: _source/docs/api/resources/idps.md (hunk 2 above)
Comment (mystiberry-okta) [inline _source/docs/api/resources/idps.md]: Will move the reference as requested.

File: _source/docs/api/resources/idps.md [hunk 3]
```diff
@@ -1016,9 +1016,12 @@ Specifies the behavior for establishing, validating, and matching a username for
...
 | filter           | Optional [regular expression pattern](https://en.wikipedia.org/wiki/Regular_expression) used to filter untrusted IdP usernames      | String                                                 | TRUE     | FALSE    | 0         | 1024      |                                                                     |
-| matchType        | Determines the Okta user profile attribute match conditions for account linking and authentication of the transformed IdP username  | `USERNAME`, `EMAIL`, or `USERNAME_OR_EMAIL`            | FALSE    | FALSE    |           |           |                                                                     |
+| matchType        | Determines the Okta user profile attribute match conditions for account linking and authentication of the transformed IdP username  | `USERNAME`, `EMAIL`, `USERNAME_OR_EMAIL` or any valid Okta user profile attribute (for example, 'firstName')*.      | FALSE    | FALSE    |           |           |                                                                     |
```
Comment (cwu-okta) [inline _source/docs/api/resources/idps.md]: I don't think we need the 'for example'

File: _source/docs/api/resources/idps.md [hunk 4]
```diff
@@ -1016,9 +1016,12 @@ Specifies the behavior for establishing, validating, and matching a username for
...
+| matchType        | Determines the Okta user profile attribute match conditions for account linking and authentication of the transformed IdP username  | `USERNAME`, `EMAIL`, `USERNAME_OR_EMAIL` or any valid Okta user profile attribute*.      | FALSE    | FALSE    |           |           |                                                                     |
 |------------------+-------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------+----------+----------+-----------+-----------+---------------------------------------------------------------------|
 
+* Inbound SAML Extensible User Matching is required to use Okta user profile attributes for `matchType`. We currently do not support USERNAME_OR_EMAIL for  Extensible User Matching. It's is an Early Access feature which you can request from Okta Support.
```
Comment (larsjohansen-okta) [inline _source/docs/api/resources/idps.md]: There's double spacing here ```...USERNAME_OR_EMAIL for  Extensib...```, and an unnecessary ```is``` here: ```...It's is an...```. I do think that ```Inbound SAML Extensible User Matching is required...``` sounds a bit weird, I fear it is not obvious to the reader that Inbound SAML Extensible User Matching is a feature.

File: _source/docs/api/resources/idps.md (hunk 3 above)
Comment (ashishgupta-okta) [inline _source/docs/api/resources/idps.md]: removed

File: _source/docs/api/resources/idps.md (hunk 4 above)
Comment (ashishgupta-okta) [inline _source/docs/api/resources/idps.md]: hmm. @mystiberry-okta can give better suggestions

**Comment by taylorlaubach-okta** [comment]
In the future, it would be easier to discuss the needed text in person or over hipchat before getting to the point of a PR. I don't have enough context to really evaluate this from a UX perspective.

That said, I'm finding this text to be confusing:

"IdP Extensible User Matching is required to use Okta user profile attributes for `matchType`. We currently do not support USERNAME_OR_EMAIL for Extensible User Matching. It's an Early Access feature which you can request from Okta Support."

If I'm understanding the text correctly, I'd suggest:

'To use Okta user profile attributes for `matchType`, you must request the "IdP Extensible User Matching" Feature, which is currently in EA. Talk to Okta Support to have this enabled.'

I'm not sure that we need to list 'matchType' in the text above. Is 'matchType' written in the UI where this appears? A screenshot would be helpful here.

Also, I'm not sure why we list  "username or email" in the first set of updated text, only to follow up by saying we don't support it.

**Comment by ashishgupta-okta** [comment]
This mostly requires a review from @mystiberry-okta . This is mostly for our IDP APIs developer doc http://developer.okta.com/docs/api/resources/idps.html#identity-provider-model. The matchType is part of the API request object. 'username or email'  matchType is supported without Elastic Search but not with Elastic Search.

**Comment by mystiberry-okta** [comment]
I'm just trying to get a simpler name for the feature. Hang on...

**Comment by mystiberry-okta** [comment]
I'm asking Hassen about a slight tweak to the feature name, otherwise it looks good. 
@taylorlaubach-okta 's right, stating the goal before the requirement is a doc best practice (my bad!). I can fix that.

**Comment by taylorlaubach-okta** [comment]
Sounds good. Sorry, I misunderstood. If this has no product UI impact, I defer to @mystiberry-okta. Mysti, if you have anything specific you'd like to get another set of eyes on, let me know. :)

**Comment by mystiberry-okta** [comment]
Looks good to me!

**Comment by mystiberry-okta** [comment]
Waiting for https://oktainc.atlassian.net/browse/OKTA-93061 to reach done status, also, no change on the feature name.

**Comment by ashishgupta-okta** [comment]
There are more changes coming @mystiberry-okta

File: _source/docs/api/resources/idps.md [hunk 5]
```diff
@@ -1016,9 +1016,13 @@ Specifies the behavior for establishing, validating, and matching a username for
...
+| matchType        | Determines the Okta user profile attribute match conditions for account linking and authentication of the transformed IdP username  | `USERNAME`, `EMAIL`, `USERNAME_OR_EMAIL` or `CUSTOM_ATTRIBUTE`      | FALSE    | FALSE    |           |           |  
+| matchAttribute        | Okta user base or custom profile attribute for matching transformed IdP username. Only for matchType `CUSTOM_ATTRIBUTE` and `SAML2` IdP*  | String      | TRUE    | FALSE    |           |           | Must be a valid Okta user profile attribute                                                                  |
 |------------------+-------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------+----------+----------+-----------+-----------+---------------------------------------------------------------------|
 
+* To use 'CUSTOM_ATTRIBUTE' for matchType, you must request the "IdP Extensible User Matching" Feature, which is currently in EA. Talk to Okta Support to have this enabled.
```
Comment (cwu-okta) [inline _source/docs/api/resources/idps.md]: Spoke to Hassen and we are offically calling this enhanced Inbound SAML (or similar.) Please change text to say something of a similar nature. One suggestion is ```you must request enhanced Inbound SAML, which is currently in EA...```

**Comment by mystiberry-okta** [comment]
Ack on all comments. Will work on this tomorrow (Sat) so you can review Monday.

**Comment by ashishgupta-okta** [comment]
image

**Comment by ashishgupta-okta** [comment]
image

**Comment by larsjohansen-okta** [comment]
🚢

**Comment by mystiberry-okta** [comment]
This is on my list to merge for next Wed's dev doc site refresh.

**Comment by mystiberry-okta** [comment]
@ashishgupta-okta , I removed the feature name since there is no feature name. Please review.

**Comment by ashishgupta-okta** [comment]
@mystiberry-okta I thought we were calling it "Enhanced Inbound SAML" ? @cwu-okta @larsjohansen-okta

**Comment by mystiberry-okta** [comment]
I can put that back--am I misremembering what was said? Question about the latency caveat still stands though.

**Comment by ashishgupta-okta** [comment]
Also , we have handled the latency issues in the code to an extent. We lookup Ids from Elastic Search and then fetch latest data about user from DB so I don't think we need that.

**Comment by mystiberry-okta** [comment]
So, is there an easy way to describe which attributes work as a value for CUSTOM_ATTRIBUTE? It's a subset of all base profile attributes, but I don't know which ones. And I'm not sure if all custom attributes are available (is there any dependency on field type or...?) @ashishgupta-okta

**Comment by ashishgupta-okta** [comment]
@mystiberry-okta we only support String,Integer and Number type so we can mention that "Must be a valid Okta user profile attribute  of type String, Integer or Number"

**Pull request merged**